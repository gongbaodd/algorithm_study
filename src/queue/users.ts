import { Queue } from "../queue/lib";
import { LNode } from "../linked_node/lib";

interface User {
  id: number;
  index: number;
  name: string;
}

export class UserQueue extends Queue<User> {
  private updateIndex() {
    if (!this.pHead?.next) {
      return;
    }

    let i = 0;
    let current = this.pHead;
    do {
      (current.data as User).index = i;
      i += 1;
      current = current.next as LNode<User>;
    } while (current);
  }

  deQueue() {
    super.deQueue();
    this.updateIndex();
  }

  enQueue(user: User) {
    super.enQueue(user);
    this.updateIndex();
  }
}
