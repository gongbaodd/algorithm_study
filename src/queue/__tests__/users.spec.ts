import { UserQueue } from "../users";

describe("A user queue system", () => {
  describe("add three users", () => {
    it("should give the index", () => {
      const queue = new UserQueue();

      queue.enQueue({ name: "a", index: 0, id: 1 });
      queue.enQueue({ name: "b", index: 0, id: 2 });
      queue.enQueue({ name: "c", index: 0, id: 3 });

      expect(queue.size()).toEqual(3);
      expect(queue).toMatchSnapshot();

      queue.deQueue();

      expect(queue.size()).toEqual(2);
      expect(queue).toMatchSnapshot();
    });
  });
});
