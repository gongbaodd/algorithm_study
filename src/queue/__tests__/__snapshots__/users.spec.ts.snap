// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A user queue system add three users should give the index 1`] = `
UserQueue {
  "pEnd": LNode {
    "data": Object {
      "id": 3,
      "index": 2,
      "name": "c",
    },
    "next": null,
  },
  "pHead": LNode {
    "data": Object {
      "id": 1,
      "index": 0,
      "name": "a",
    },
    "next": LNode {
      "data": Object {
        "id": 2,
        "index": 1,
        "name": "b",
      },
      "next": LNode {
        "data": Object {
          "id": 3,
          "index": 2,
          "name": "c",
        },
        "next": null,
      },
    },
  },
}
`;

exports[`A user queue system add three users should give the index 2`] = `
UserQueue {
  "pEnd": LNode {
    "data": Object {
      "id": 3,
      "index": 1,
      "name": "c",
    },
    "next": null,
  },
  "pHead": LNode {
    "data": Object {
      "id": 2,
      "index": 0,
      "name": "b",
    },
    "next": LNode {
      "data": Object {
        "id": 3,
        "index": 1,
        "name": "c",
      },
      "next": null,
    },
  },
}
`;
